// @redwoodjs/cli/dist/commands/dbCommands/introspect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = void 0;

var options = _interopRequireWildcard(require("./options"));

var _lib = require("../../lib");

const command = 'introspect';
exports.command = command;
const description = 'Introspect your database and generate models in ./api/prisma/schema.prisma, overwriting existing models';
exports.description = description;

const builder = yargs => {
  yargs.option('verbose', options.verbose()).option('schema', options.schema()).epilogue(options.epilogue());
};

exports.builder = builder;

const handler = async ({
  verbose = true,
  schema
}) => {
  return await (0, _lib.runCommandTask)([{
    title: 'Introspecting your database...',
    cmd: 'yarn prisma',
    args: ['introspect', schema && `--schema="${schema}"`],
    opts: {
      cwd: (0, _lib.getPaths)().api.db
    }
  }], {
    verbose
  });
};

exports.handler = handler;
