// @redwoodjs/cli/dist/commands/dbCommands/save.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = void 0;

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/filter"));

var options = _interopRequireWildcard(require("./options"));

var _lib = require("../../lib");

const command = 'save [name..]';
exports.command = command;
const description = 'Create a new migration';
exports.description = description;

const builder = yargs => {
  yargs.positional('name', {
    default: 'migration',
    description: 'Name of the migration',
    type: 'string'
  }).option('verbose', options.verbose()).option('schema', options.schema()).epilogue(options.epilogue());
};

exports.builder = builder;

const handler = async ({
  name = 'migration',
  verbose = true,
  schema
}) => {
  var _context;

  await (0, _lib.runCommandTask)([{
    title: 'Creating database migration...',
    cmd: 'yarn prisma',
    args: (0, _filter.default)(_context = ['migrate save', name.length && `--name "${name}"`, '--create-db', '--experimental', schema && `--schema="${schema}"`]).call(_context, Boolean)
  }], {
    verbose
  });
};

exports.handler = handler;
