// @redwoodjs/cli/dist/commands/dbCommands/seed.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = void 0;

var _terminalLink = _interopRequireDefault(require("terminal-link"));

var _lib = require("../../lib");

const command = 'seed';
exports.command = command;
const description = 'Seed your database with test data';
exports.description = description;

const builder = yargs => {
  yargs.epilogue(`Also see the ${(0, _terminalLink.default)('Redwood CLI Reference', 'https://redwoodjs.com/reference/command-line-interface#db-seed')}`);
};

exports.builder = builder;

const handler = () => {
  (0, _lib.runCommandTask)([{
    title: 'Seeding your database...',
    cmd: 'node',
    args: ['seeds.js'],
    opts: {
      cwd: (0, _lib.getPaths)().api.db
    }
  }], {
    verbose: true
  });
};

exports.handler = handler;
