// @redwoodjs/cli/dist/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.getDefaultArgs = exports.runCommandTask = exports.removeRoutesFromRouterTask = exports.addRoutesToRouterTask = exports.cleanupEmptyDirsTask = exports.deleteFilesTask = exports.writeFilesTask = exports.transformTSToJS = exports.prettierOptions = exports.getPaths = exports.bytes = exports.writeFile = exports.deleteFile = exports.readFile = exports.prettify = exports.generateTemplate = exports.templateRoot = exports.nameVariants = exports.getSchemaDefinitions = exports.getEnum = exports.getSchema = exports.asyncForEach = void 0;

var _entries = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/object/entries"));

var _reverse = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reverse"));

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reduce"));

var _set = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/set"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/object/keys"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/map"));

var _some = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/some"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/for-each"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/includes"));

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/find"));

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var babel = _interopRequireWildcard(require("@babel/core"));

var _sdk = require("@prisma/sdk");

var _camelcase = _interopRequireDefault(require("camelcase"));

var _decamelize = _interopRequireDefault(require("decamelize"));

var _execa = _interopRequireDefault(require("execa"));

var _listr = _interopRequireDefault(require("listr"));

var _listrVerboseRenderer = _interopRequireDefault(require("listr-verbose-renderer"));

var _string = _interopRequireDefault(require("lodash/string"));

var _paramCase = require("param-case");

var _pascalcase = _interopRequireDefault(require("pascalcase"));

var _pluralize = _interopRequireDefault(require("pluralize"));

var _prettier = require("prettier");

var _internal = require("@redwoodjs/internal");

var _colors = _interopRequireDefault(require("./colors"));

const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};
/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target application. If no `name` is given then the
 * entire schema is returned.
 */


exports.asyncForEach = asyncForEach;

const getSchema = async name => {
  const schema = await getSchemaDefinitions();

  if (name) {
    var _context;

    const model = (0, _find.default)(_context = schema.datamodel.models).call(_context, model => {
      return model.name === name;
    });

    if (model) {
      return model;
    } else {
      throw new Error(`No schema definition found for \`${name}\` in schema.prisma file`);
    }
  }

  return schema.metadata.datamodel;
};
/**
 * Returns the enum defined with the given `name` parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * all enum definitions are returned
 */


exports.getSchema = getSchema;

const getEnum = async name => {
  const schema = await getSchemaDefinitions();

  if (name) {
    var _context2;

    const model = (0, _find.default)(_context2 = schema.datamodel.enums).call(_context2, model => {
      return model.name === name;
    });

    if (model) {
      return model;
    } else {
      throw new Error(`No enum schema definition found for \`${name}\` in schema.prisma file`);
    }
  }

  return schema.metadata.datamodel.enums;
};
/*
 * Returns the DMMF defined by `prisma` resolving the relevant `schema.prisma` path.
 */


exports.getEnum = getEnum;

const getSchemaDefinitions = async () => {
  const schemaPath = _path.default.join(getPaths().api.db, 'schema.prisma');

  const metadata = await (0, _sdk.getDMMF)({
    datamodel: readFile(schemaPath.toString())
  });
  return metadata;
};
/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
 * singularConstantName: FOO_BAR
 * pluralConstantName: FOO_BARS
*/


exports.getSchemaDefinitions = getSchemaDefinitions;

const nameVariants = name => {
  const normalizedName = (0, _pascalcase.default)((0, _paramCase.paramCase)(_pluralize.default.singular(name)));
  return {
    pascalName: (0, _pascalcase.default)(name),
    camelName: (0, _camelcase.default)(name),
    singularPascalName: normalizedName,
    pluralPascalName: (0, _pluralize.default)(normalizedName),
    singularCamelName: (0, _camelcase.default)(normalizedName),
    pluralCamelName: (0, _camelcase.default)((0, _pluralize.default)(normalizedName)),
    singularParamName: (0, _paramCase.paramCase)(normalizedName),
    pluralParamName: (0, _paramCase.paramCase)((0, _pluralize.default)(normalizedName)),
    singularConstantName: (0, _decamelize.default)(normalizedName).toUpperCase(),
    pluralConstantName: (0, _decamelize.default)((0, _pluralize.default)(normalizedName)).toUpperCase()
  };
};

exports.nameVariants = nameVariants;

const templateRoot = _path.default.resolve(__dirname, '../commands/generate');

exports.templateRoot = templateRoot;

const generateTemplate = (templateFilename, {
  name,
  root,
  ...rest
}) => {
  const templatePath = _path.default.join(root || templateRoot, templateFilename);

  const template = _string.default.template(readFile(templatePath).toString());

  const renderedTemplate = template({
    name,
    ...nameVariants(name),
    ...rest
  });
  return prettify(templateFilename, renderedTemplate);
};

exports.generateTemplate = generateTemplate;

const prettify = (templateFilename, renderedTemplate) => {
  // We format .js and .css templates, we need to tell prettier which parser
  // we're using.
  // https://prettier.io/docs/en/options.html#parser
  const parser = {
    '.css': 'css',
    '.js': 'babel',
    '.ts': 'babel-ts'
  }[_path.default.extname(templateFilename.replace('.template', ''))];

  if (typeof parser === 'undefined') {
    return renderedTemplate;
  }

  return (0, _prettier.format)(renderedTemplate, { ...prettierOptions(),
    parser
  });
};

exports.prettify = prettify;

const readFile = target => _fs.default.readFileSync(target);

exports.readFile = readFile;
const SUPPORTED_EXTENSIONS = ['.js', '.ts', '.tsx'];

const deleteFile = file => {
  const extension = _path.default.extname(file);

  if ((0, _includes.default)(SUPPORTED_EXTENSIONS).call(SUPPORTED_EXTENSIONS, extension)) {
    const baseFile = getBaseFile(file);
    (0, _forEach.default)(SUPPORTED_EXTENSIONS).call(SUPPORTED_EXTENSIONS, ext => {
      const f = baseFile + ext;

      if (_fs.default.existsSync(f)) {
        _fs.default.unlinkSync(f);
      }
    });
  } else {
    _fs.default.unlinkSync(file);
  }
};

exports.deleteFile = deleteFile;

const getBaseFile = file => file.replace(/\.\w*$/, '');

const existsAnyExtensionSync = file => {
  const extension = _path.default.extname(file);

  if ((0, _includes.default)(SUPPORTED_EXTENSIONS).call(SUPPORTED_EXTENSIONS, extension)) {
    const baseFile = getBaseFile(file);
    return (0, _some.default)(SUPPORTED_EXTENSIONS).call(SUPPORTED_EXTENSIONS, ext => _fs.default.existsSync(baseFile + ext));
  }

  return _fs.default.existsSync(file);
};

const writeFile = (target, contents, {
  overwriteExisting = false
} = {}, task = {}) => {
  const {
    base
  } = getPaths();
  task.title = `Writing \`./${_path.default.relative(base, target)}\``;

  if (!overwriteExisting && _fs.default.existsSync(target)) {
    throw new Error(`${target} already exists.`);
  }

  const filename = _path.default.basename(target);

  const targetDir = target.replace(filename, '');

  _fs.default.mkdirSync(targetDir, {
    recursive: true
  });

  _fs.default.writeFileSync(target, contents);

  task.title = `Successfully wrote file \`./${_path.default.relative(base, target)}\``;
};

exports.writeFile = writeFile;

const bytes = contents => Buffer.byteLength(contents, 'utf8');
/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */


exports.bytes = bytes;

const getPaths = () => {
  try {
    return (0, _internal.getPaths)();
  } catch (e) {
    console.error(_colors.default.error(e.message));
    process.exit(0);
  }
};
/**
 * This returns the config present in `prettier.config.js` of a Redwood project.
 */


exports.getPaths = getPaths;

const prettierOptions = () => {
  try {
    return require(_path.default.join(getPaths().base, 'prettier.config.js'));
  } catch (e) {
    return undefined;
  }
}; // TODO: Move this into `generateTemplate` when all templates have TS support

/*
 * Convert a generated TS template file into JS.
 */


exports.prettierOptions = prettierOptions;

const transformTSToJS = (filename, content) => {
  const {
    code
  } = babel.transform(content, {
    filename,
    // If you ran `yarn rw generate` in `./web` transformSync would import the `.babelrc.js` file,
    // in `./web`? despite us setting `configFile: false`.
    cwd: process.env.NODE_ENV === 'test' ? undefined : getPaths().base,
    configFile: false,
    plugins: [['@babel/plugin-transform-typescript', {
      isTSX: true,
      allExtensions: true
    }]],
    retainLines: true
  });
  return prettify(filename.replace(/\.ts$/, '.js'), code);
};
/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */


exports.transformTSToJS = transformTSToJS;

const writeFilesTask = (files, options) => {
  var _context3;

  const {
    base
  } = getPaths();
  return new _listr.default((0, _map.default)(_context3 = (0, _keys.default)(files)).call(_context3, file => {
    const contents = files[file];
    return {
      title: `...waiting to write file \`./${_path.default.relative(base, file)}\`...`,
      task: (ctx, task) => writeFile(file, contents, options, task)
    };
  }));
};
/**
 * Creates a list of tasks that delete files from the disk.
 *
 * @param files - {[filepath]: contents}
 */


exports.writeFilesTask = writeFilesTask;

const deleteFilesTask = files => {
  var _context4;

  const {
    base
  } = getPaths();
  return new _listr.default([...(0, _map.default)(_context4 = (0, _keys.default)(files)).call(_context4, file => {
    return {
      title: `Destroying \`./${_path.default.relative(base, getBaseFile(file))}\`...`,
      skip: () => !existsAnyExtensionSync(file) && `File doesn't exist`,
      task: () => deleteFile(file)
    };
  }), {
    title: 'Cleaning up empty directories...',
    task: () => cleanupEmptyDirsTask(files)
  }]);
};
/**
 * @param files - {[filepath]: contents}
 */


exports.deleteFilesTask = deleteFilesTask;

const cleanupEmptyDirsTask = files => {
  var _context5;

  const {
    base
  } = getPaths();
  const allDirs = (0, _map.default)(_context5 = (0, _keys.default)(files)).call(_context5, file => _path.default.dirname(file));
  const uniqueDirs = [...new _set.default(allDirs)];
  return new _listr.default((0, _map.default)(uniqueDirs).call(uniqueDirs, dir => {
    return {
      title: `Removing empty \`./${_path.default.relative(base, dir)}\`...`,
      task: () => _fs.default.rmdirSync(dir),
      skip: () => {
        if (!_fs.default.existsSync(dir)) {
          return `Doesn't exist`;
        }

        if (_fs.default.readdirSync(dir).length > 0) {
          return 'Not empty';
        }

        return false;
      }
    };
  }));
};
/**
 * Update the project's routes file.
 */


exports.cleanupEmptyDirsTask = cleanupEmptyDirsTask;

const addRoutesToRouterTask = routes => {
  var _context6;

  const redwoodPaths = getPaths();
  const routesContent = readFile(redwoodPaths.web.routes).toString();
  const newRoutesContent = (0, _reduce.default)(_context6 = (0, _reverse.default)(routes).call(routes)).call(_context6, (content, route) => {
    if ((0, _includes.default)(content).call(content, route)) {
      return content;
    }

    return content.replace(/<Router>(\s*)/, `<Router>$1${route}$1`);
  }, routesContent);
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true
  });
};
/**
 * Remove named routes from the project's routes file.
 *
 * @param {string[]} routes - Route names
 */


exports.addRoutesToRouterTask = addRoutesToRouterTask;

const removeRoutesFromRouterTask = routes => {
  const redwoodPaths = getPaths();
  const routesContent = readFile(redwoodPaths.web.routes).toString();
  const newRoutesContent = (0, _reduce.default)(routes).call(routes, (content, route) => {
    const matchRouteByName = new RegExp(`\\s*<Route[^>]*name="${route}"[^>]*/>`);
    return content.replace(matchRouteByName, '');
  }, routesContent);
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true
  });
};

exports.removeRoutesFromRouterTask = removeRoutesFromRouterTask;

const runCommandTask = async (commands, {
  verbose
}) => {
  const tasks = new _listr.default((0, _map.default)(commands).call(commands, ({
    title,
    cmd,
    args,
    opts = {}
  }) => ({
    title,
    task: async () => {
      return (0, _execa.default)(cmd, args, {
        shell: true,
        cwd: `${getPaths().base}/api`,
        stdio: verbose ? 'inherit' : 'pipe',
        extendEnv: true,
        cleanup: true,
        ...opts
      });
    }
  })), {
    renderer: verbose && _listrVerboseRenderer.default,
    dateFormat: false
  });

  try {
    await tasks.run();
    return true;
  } catch (e) {
    console.log(_colors.default.error(e.message));
    return false;
  }
};
/*
 * Extract default CLI args from an exported builder
 */


exports.runCommandTask = runCommandTask;

const getDefaultArgs = builder => {
  var _context7;

  return (0, _reduce.default)(_context7 = (0, _entries.default)(builder)).call(_context7, (agg, [k, v]) => {
    agg[k] = v.default;
    return agg;
  }, {});
};

exports.getDefaultArgs = getDefaultArgs;
