// @redwoodjs/cli/dist/commands/generate/auth/providers/supabase.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.notes = exports.apiPackages = exports.webPackages = exports.config = void 0;
// the lines that need to be added to index.js
const config = {
  imports: [`import { createClient } from '@supabase/supabase-js'`],
  init: `const supabaseClient = createClient(
    process.env.SUPABASE_URL,
    process.env.SUPABASE_KEY
  )`,
  authProvider: {
    client: 'supabaseClient',
    type: 'supabase'
  }
}; // required packages to install

exports.config = config;
const webPackages = ['@supabase/supabase-js'];
exports.webPackages = webPackages;
const apiPackages = []; // any notes to print out when the job is done

exports.apiPackages = apiPackages;
const notes = ['You will need to add your Supabase URL and key to your .env file.', 'See: https://supabase.io/docs/library/getting-started#reference'];
exports.notes = notes;
