// @redwoodjs/auth/dist/authClients/auth0.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.auth0 = void 0;

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/includes"));

// TODO: Map out this user properly.
const auth0 = client => {
  return {
    type: 'auth0',
    client,
    restoreAuthState: async () => {
      var _context;

      if ((0, _includes.default)(_context = window.location.search).call(_context, 'code=')) {
        const {
          appState
        } = await client.handleRedirectCallback();
        window.history.replaceState({}, document.title, appState && appState.targetUrl ? appState.targetUrl : window.location.pathname);
      }
    },
    login: async options => client.loginWithRedirect(options),
    logout: options => client.logout(options),
    signup: async (options) => client.loginWithRedirect({ ...options,
      screen_hint: 'signup',
      prompt: 'login'
    }),
    getToken: async () => client.getTokenSilently(),
    getUserMetadata: async () => {
      const user = await client.getUser();
      return user || null;
    }
  };
};

exports.auth0 = auth0;
