// @redwoodjs/router/dist/location.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.useLocation = exports.LocationContext = exports.LocationProvider = exports.Location = void 0;

var _react = _interopRequireDefault(require("react"));

var _internal = require("./internal");

const LocationContext = (0, _internal.createNamedContext)('Location');
exports.LocationContext = LocationContext;

class LocationProvider extends _react.default.Component {
  constructor(...args) {
    super(...args);
    this.HISTORY_LISTENER_ID = undefined;
    this.state = {
      context: this.getContext()
    };
  }

  getContext() {
    const {
      pathname,
      search,
      hash
    } = this.props.location;
    return {
      pathname,
      search,
      hash
    };
  }

  componentDidMount() {
    this.HISTORY_LISTENER_ID = _internal.gHistory.listen(() => {
      this.setState(() => ({
        context: this.getContext()
      }));
    });
  }

  componentWillUnmount() {
    _internal.gHistory.remove(this.HISTORY_LISTENER_ID);
  }

  render() {
    let {
      children
    } = this.props;
    let {
      context
    } = this.state;
    return /*#__PURE__*/_react.default.createElement(LocationContext.Provider, {
      value: context
    }, typeof children === 'function' ? children(context) : children || null);
  }

}

exports.LocationProvider = LocationProvider;
LocationProvider.defaultProps = {
  location: window.location
};

const Location = ({
  children
}) => /*#__PURE__*/_react.default.createElement(LocationContext.Consumer, null, context => context ? children(context) : /*#__PURE__*/_react.default.createElement(LocationProvider, null, children));

exports.Location = Location;

const useLocation = () => {
  const location = _react.default.useContext(LocationContext);

  return location;
};

exports.useLocation = useLocation;
