// @redwoodjs/web/dist/flash/FlashContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.useFlash = exports.FlashProvider = exports.FlashContext = void 0;

var _react = _interopRequireWildcard(require("react"));

var _FlashReducer = _interopRequireDefault(require("./FlashReducer"));

// initial state
const initialState = {
  messages: []
}; // create context

const FlashContext = /*#__PURE__*/(0, _react.createContext)(initialState); // providor component

exports.FlashContext = FlashContext;

const FlashProvider = ({
  children
}) => {
  const [state, dispatch] = (0, _react.useReducer)(_FlashReducer.default, initialState); // dispatch actions to reducer

  function addMessage(text, options) {
    const message = {
      text,
      ...options
    };
    dispatch({
      type: 'ADD_MESSAGE',
      payload: message
    });
  }

  function dismissMessage(messageId) {
    dispatch({
      type: 'DISMISS_MESSAGE',
      payload: messageId
    });
  }

  function cycleMessage(messageId) {
    dispatch({
      type: 'CYCLE_MESSAGE',
      payload: messageId
    });
  } // render


  return /*#__PURE__*/_react.default.createElement(FlashContext.Provider, {
    value: {
      messages: state.messages,
      addMessage,
      dismissMessage,
      cycleMessage
    }
  }, children);
}; // a hook to use flash values


exports.FlashProvider = FlashProvider;

const useFlash = () => {
  const flash = _react.default.useContext(FlashContext);

  return flash;
};

exports.useFlash = useFlash;
